Do not ICE when trying to get layout of an unexpected type

gurry opened the pull request:
Fixes #126942

See issue for why the ICE occurs.

In this PR we return an error from layout query instead of raising a bug when we encounter a type that we cannot compute the layout of.

**Edit**

Put the fix in `try_normalize_erasing_regions` instead of `layout_of_uncached` because we were triggering a debug_assertion in the former. 

Also fixes #127804

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9953836116/job/27498202408) [(plain)](https://github.com/rust-lang/rust/commit/48ae9e6f73e1f790c91f96348aa36bdd3cff6b7c/checks/27498202408/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
failures:

---- [ui] tests/ui/layout/ice-layout-of-invalid-type-126942.rs stdout ----

error: Error: expected failure status (Some(1)) but received status Some(101).
status: exit status: 101
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/layout/ice-layout-of-invalid-type-126942.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/layout/ice-layout-of-invalid-type-126942" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/layout/ice-layout-of-invalid-type-126942/auxiliary"
--- stderr -------------------------------
error[E0207]: the type parameter `T` is not constrained by the impl trait, self type, or predicates
##[error]  --> /checkout/tests/ui/layout/ice-layout-of-invalid-type-126942.rs:12:6
   |
   |
LL | impl<T: ?Sized> Every for Thing {

error[E0283]: type annotations needed
##[error]  --> /checkout/tests/ui/layout/ice-layout-of-invalid-type-126942.rs:19:11
   |
   |
LL | static I: <Thing as Every>::Assoc = 3;
   |           ^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `T`
   = note: cannot satisfy `_: Sync`
   = note: shared static variables must have a type that implements `Sync`

thread 'rustc' panicked at /checkout/compiler/rustc_traits/src/normalize_erasing_regions.rs:45:13:
thread 'rustc' panicked at /checkout/compiler/rustc_traits/src/normalize_erasing_regions.rs:45:13:
?0t
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: rustc_traits::normalize_erasing_regions::try_normalize_after_erasing_regions::<rustc_middle::ty::generic_args::GenericArg>
   3: <rustc_traits::normalize_erasing_regions::provide::{closure#0} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, rustc_middle::ty::ParamEnvAnd<rustc_middle::ty::generic_args::GenericArg>)>>::call_once
      [... omitted 2 frames ...]
   4: rustc_middle::query::plumbing::query_get_at::<rustc_query_system::query::caches::DefaultCache<rustc_middle::ty::ParamEnvAnd<rustc_middle::ty::generic_args::GenericArg>, rustc_middle::query::erase::Erased<[u8; 8]>>>
   5: <rustc_middle::ty::normalize_erasing_regions::TryNormalizeAfterErasingRegionsFolder>::try_normalize_generic_arg_after_erasing_regions
   6: <rustc_middle::ty::normalize_erasing_regions::TryNormalizeAfterErasingRegionsFolder as rustc_type_ir::fold::FallibleTypeFolder<rustc_middle::ty::context::TyCtxt>>::try_fold_ty
   8: rustc_ty_utils::layout::layout_of
   8: rustc_ty_utils::layout::layout_of
      [... omitted 2 frames ...]
   9: rustc_middle::query::plumbing::query_get_at::<rustc_query_system::query::caches::DefaultCache<rustc_middle::ty::ParamEnvAnd<rustc_middle::ty::Ty>, rustc_middle::query::erase::Erased<[u8; 16]>>>
  10: rustc_hir_analysis::check::check::check_static_inhabited
  12: rustc_hir_analysis::check::wfcheck::check_well_formed
  12: rustc_hir_analysis::check::wfcheck::check_well_formed
      [... omitted 2 frames ...]
  13: <rustc_data_structures::sync::parallel::ParallelGuard>::run::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_data_structures::sync::parallel::enabled::try_par_for_each_in<&[rustc_hir::hir::ForeignItemId], rustc_span::ErrorGuaranteed, <rustc_middle::hir::ModuleItems>::par_foreign_items<rustc_hir_analysis::check::wfcheck::check_mod_type_wf::{closure#3}>::{closure#0}>::{closure#0}::{closure#0}::{closure#0}>
  14: <rustc_middle::hir::ModuleItems>::par_items::<rustc_hir_analysis::check::wfcheck::check_mod_type_wf::{closure#0}>
  15: rustc_hir_analysis::check::wfcheck::check_mod_type_wf
      [... omitted 2 frames ...]
  16: <rustc_middle::hir::map::Map>::par_for_each_module::<rustc_hir_analysis::check_crate::{closure#0}::{closure#0}>::{closure#0}
  17: <rustc_data_structures::sync::parallel::ParallelGuard>::run::<(), rustc_data_structures::sync::parallel::enabled::par_for_each_in<&rustc_hir::hir_id::OwnerId, &[rustc_hir::hir_id::OwnerId], <rustc_middle::hir::map::Map>::par_for_each_module<rustc_hir_analysis::check_crate::{closure#0}::{closure#0}>::{closure#0}>::{closure#0}::{closure#1}::{closure#0}>
  19: rustc_hir_analysis::check_crate
  20: rustc_interface::passes::analysis
  20: rustc_interface::passes::analysis
      [... omitted 2 frames ...]
  22: <rustc_interface::interface::Compiler>::enter::<rustc_driver_impl::run_compiler::{closure#0}::{closure#1}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_span::ErrorGuaranteed>>
  23: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>
  24: rustc_span::create_session_globals_then::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
---
note: please make sure that you have updated to the latest nightly

note: rustc 1.81.0-nightly (27c2f1501 2024-07-16) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0
query stack during panic:
query stack during panic:
#0 [try_normalize_generic_arg_after_erasing_regions] normalizing `<Thing as Every>::Assoc`
#1 [layout_of] computing layout of `<Thing as Every>::Assoc`
#2 [check_well_formed] checking that `I` is well-formed
#3 [check_mod_type_wf] checking that types are well-formed in top-level module
end of query stack
error: aborting due to 2 previous errors

Some errors have detailed explanations: E0207, E0283.
```

</details>

gurry commented:
@rustbot author

compiler-errors reviewed (changes requested):
Why didn't you put this check into `check_static_inhabited`? Seems like a more specific place to put it, according to the ICE message in the issue you linked.

gurry commented:
I felt that it's cleaner to put it in layout because if I put it in `check_static_inhabited` I'll have to make an extra call to`try_normalize_erasing_regions` which layout already does.

compiler-errors commented:
Well it looks like the test still ICEs anyways 🤔. I would encourage you to try to make this fix more specific if possible, to avoid changing existing assertions which were added as a sanity check.

gurry commented:
Yeah, the test is ICEing in `try_normalize_erasing_regions` and even if I make it more specific I'll still have to normalize the type and therefore it might still ICE.

It works fine on my MacBook and I don't have a Linux machine. But I guess the test's ICE is more due to x64 rather than the OS. So I'm going to test it on a Windows machine (which I do have) and see if I can fix it.

gurry commented:
I'll also try to see if I can make it more specific nonetheless.

compiler-errors commented:
This is because of a debug assertion, which you need to enable in your `config.toml`.

This has nothing to do with operating systems or CPU architectures--it's a type system bug after all.

gurry commented:
Ah, thanks for that. Saved me a lot of time.

gurry commented:
I tried to put the fix in `check_static_inhabited`, but I would have had to call the `try_normalize_erasing_regions` from there before calling `layout_of` and it is ICEing due to the `debug_assertion`. So I had make the change in `try_normalize_erasing_regions` itself to remove the `debug_assertion` and return an error instead.

If you have any suggestions make it more specific, please let me know.

gurry committed:
Fix ICE while normalizing an infer type

Returning an error instead of triggering a debug_assertion
when a normalized type still has infer

gurry commented:
@rustbot review

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9968211162/job/27543109513) [(plain)](https://github.com/rust-lang/rust/commit/9c78feccef434dbeb16541765543a6262f758d7f/checks/27543109513/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
##[group]Testing stage2 compiletest suite=crashes mode=crashes (x86_64-unknown-linux-gnu)

running 224 tests
........................................................................................  88/224
......2024-07-17T04:57:16.940333Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge."
[crashes] tests/crashes/118590.rs ... F
[crashes] tests/crashes/118590.rs ... F
...............................2024-07-17T04:57:17.199199Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge."
[crashes] tests/crashes/123141.rs ... F
................................................. 176/224
................................................. 176/224
........2024-07-17T04:57:17.676914Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge."
[crashes] tests/crashes/107362.rs ... F
.......................................

failures:
failures:

---- [crashes] tests/crashes/118590.rs stdout ----

error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge.
thread '[crashes] tests/crashes/118590.rs' panicked at src/tools/compiletest/src/runtest.rs:377:18:
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- [crashes] tests/crashes/123141.rs stdout ----


error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge.
thread '[crashes] tests/crashes/123141.rs' panicked at src/tools/compiletest/src/runtest.rs:377:18:

---- [crashes] tests/crashes/107362.rs stdout ----


error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge.
thread '[crashes] tests/crashes/107362.rs' panicked at src/tools/compiletest/src/runtest.rs:377:18:


failures:
    [crashes] tests/crashes/118590.rs
```

</details>

gurry commented:
> The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9968211162/job/27543109513) [(plain)](https://github.com/rust-lang/rust/commit/9c78feccef434dbeb16541765543a6262f758d7f/checks/27543109513/logs)
> 
> Click to see the possible cause of the failure (guessed by this bot)
> ```
>  - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
>  - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
>  - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
> ##[endgroup]
> Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
> [CI_JOB_NAME=x86_64-gnu-llvm-17]
> ---
> sccache: Starting the server...
> ##[group]Configure the build
> configure: processing command line
> configure: 
> configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
> configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
> configure: llvm.link-shared     := True
> configure: rust.thin-lto-import-instr-limit := 10
> configure: change-id            := 99999999
> ---
> ##[group]Testing stage2 compiletest suite=crashes mode=crashes (x86_64-unknown-linux-gnu)
> 
> running 224 tests
> ........................................................................................  88/224
> ......2024-07-17T04:57:16.940333Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge."
> [crashes] tests/crashes/118590.rs ... F
> [crashes] tests/crashes/118590.rs ... F
> ...............................2024-07-17T04:57:17.199199Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge."
> [crashes] tests/crashes/123141.rs ... F
> ................................................. 176/224
> ................................................. 176/224
> ........2024-07-17T04:57:17.676914Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge."
> [crashes] tests/crashes/107362.rs ... F
> .......................................
> 
> failures:
> failures:
> 
> ---- [crashes] tests/crashes/118590.rs stdout ----
> 
> error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge.
> thread '[crashes] tests/crashes/118590.rs' panicked at src/tools/compiletest/src/runtest.rs:377:18:
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> 
> ---- [crashes] tests/crashes/123141.rs stdout ----
> 
> 
> error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge.
> thread '[crashes] tests/crashes/123141.rs' panicked at src/tools/compiletest/src/runtest.rs:377:18:
> 
> ---- [crashes] tests/crashes/107362.rs stdout ----
> 
> 
> error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge.
> thread '[crashes] tests/crashes/107362.rs' panicked at src/tools/compiletest/src/runtest.rs:377:18:
> 
> 
> failures:
>     [crashes] tests/crashes/118590.rs
> ```

Woah! Didn't realise this one stone was gonna hit so many birds.

compiler-errors reviewed (changes requested):
This is not the right fix. This debug assertion exists for a good reason -- because we shouldn't be returning totally unconstrained inference variables from the `normalize_after_erasing_regions` query.

compiler-errors commented:
In general, when fixing an ICE, I would appreciate if you looker deeper than just the surface. Suppressing a debug assertion from triggering by just removing the debug assertion is not the right answer.

compiler-errors commented:
Firstly, it would be useful to understand why #126942 began to ICE in the first place. It seems like @oli-obk caused this in #121154. This is a classic case of the types of ICEs introduced with this error refactoring work, so maybe he has some solutions in mind that are less invasive than just removing debug assertions.

Perhaps one solution to this would be to project associated types to `ty::Error` if we detect an impl with unconstrained arguments. That would require querifying the `check_impl_wf` function, or at least the `enforce_impl_params_are_constrained` part. Both of those seem like better solutions than

compiler-errors commented:
Alternatively, it does seem a bit weird to me that we're checking whether static layouts are inhabited this early in compilation (`check_mod_wf`). I wonder if we can move this check around to somewhere else to avoid calling it when the impls aren't even WF.

gurry commented:
Awesome. These are all good suggestions. I'll investigate further.

Thanks :).

gurry commented:
> This is not the right fix. This debug assertion exists for a good reason -- because we shouldn't be returning totally unconstrained inference variables from the `normalize_after_erasing_regions` query.

After this change we not returning inference variables. We're returning an error. So it's not all bad. But still it's not the best solution.

oli-obk commented:
Unfortunately, as with most wfcheck logic, we can't invoke it from anywhere else but top-level queries, as it invokes so many other type queries, that we run into cycles left and right. I did assume `enforce_impl_params_are_constrained` on its own could be small enough to allow us to call it from other queries, but it turns out it does some stuff for backwards compatibility that makes that hard: https://rust-lang.zulipchat.com/#narrow/stream/144729-t-types/topic/unconstrained.20lifetimes.20in.20impls

gurry commented:
Closing in favour of #127973
